File: RetosOtherLanguages/RetoN°4/RetoN°4.pl
use strict;
use warnings;

sub decode {
    my ($message) = @_;

    my @stack;

    for my $char (split //, $message) {
        if ($char eq ')') {
            my $reversed = '';
            while (@stack && $stack[-1] ne '(') {
                $reversed .= pop @stack;
            }
            # Remove the '(' from stack
            pop @stack;
            # Push the reversed string back into stack
            push @stack, split //, $reversed;
        } else {
            push @stack, $char;
        }
    }

    return join('', @stack);
}

my $a = decode("hola (odnum)");
print "$a\n"; # Hola mundo

my $b = decode("(olleh) (dlrow)!");
print "$b\n"; # Hello World!

my $c = decode("sa(u(cla)atn)s");
print "$c\n"; # santaclaus

my $d = decode("((nta)(sa))");
print "$d\n"; # Hello World!

File: RetosOtherLanguages/RetoN°4/RetoN°4.lua
function decode(message)
    local stack = {}
    local result = {}

    for i = 1, #message do
        local c = message:sub(i, i)

        if c == ')' then
            local reversed = {}
            while #stack > 0 and stack[#stack] ~= '(' do
                table.insert(reversed, table.remove(stack))
            end
            -- Eliminar el '(' de la pila
            table.remove(stack)
            -- Insertar la cadena invertida de nuevo en la pila
            for j = 1, #reversed do
                table.insert(stack, reversed[j])
            end
        else
            table.insert(stack, c)
        end
    end

    for i = 1, #stack do
        table.insert(result, stack[i])
    end

    return table.concat(result)
end

local a = decode("hola (odnum)")
print(a)  -- Hola mundo

local b = decode("(olleh) (dlrow)!")
print(b)  -- Hello World!

local c = decode("sa(u(cla)atn)s")
print(c)  -- santaclaus

local d = decode("((nta)(sa))")
print(d)  -- Hello World!
